sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala

```diff
@@ -112,34 +108,36 @@ private[smartdatalake] case class ActionDAGRun(dag: DAG[Action], executionId: SD
     result.filter(_.isSuccess).map(_.get)
   }
 
-  def prepare(implicit context: ActionPipelineContext): Unit = {
-    context.phase = ExecutionPhase.Prepare
+  def prepare(context: ActionPipelineContext): Unit = {
+    implicit val phaseContext: ActionPipelineContext = context.copy(phase = ExecutionPhase.Prepare)
     // run prepare for every node
-    run[DummyDAGResult](context.phase) {
+    run[DummyDAGResult](phaseContext.phase) {
       case (node: InitDAGNode, _) =>
         node.edges.map(dataObjectId => DummyDAGResult(dataObjectId))
       case (node: Action, _) =>
-        node.prepare
+        val actionContext = phaseContext.copy(currentAction = Some(node))
+        node.prepare(actionContext)
         node.outputs.map(outputDO => DummyDAGResult(outputDO.id.id))
       case x => throw new IllegalStateException(s"Unmatched case $x")
     }
   }
 
-  def init(implicit context: ActionPipelineContext): Seq[SubFeed] = {
-    context.phase = ExecutionPhase.Init
+  def init(context: ActionPipelineContext): Seq[SubFeed] = {
+    implicit val phaseContext: ActionPipelineContext = context.copy(phase = ExecutionPhase.Init)
```

Comment:
how can it happen that context.phase is not Init in this situation? seems strange

Reply:
Well, this is the beginning of the Init-Phase and we have to initialize the context correctly.

Reply:
And why not directly change the original context variable when you call the respective function ?, eg
actionDAGRun.init(context.copy(phase = ExecutionPhase.Init))

Reply:
Yes, it's called in many places. Like this you don't have to remember to pass a context with phase = ExecutionPhase.Init to this function.